Apply Weight Window Upon Surface Crossing

XinyanBradley opened the pull request:
<!--
If you are a first-time contributor to OpenMC, please have a look at our
contributing guidelines:
https://github.com/openmc-dev/openmc/blob/develop/CONTRIBUTING.md
-->

# Description

Apply weight window split/roulette upon surface crossing.

Fixes # (issue)

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have run [clang-format](https://docs.openmc.org/en/latest/devguide/styleguide.html#automatic-formatting) on any C++ source files (if applicable)
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)
<!--
While tests will automatically be checked by CI, it is good practice to
ensure that they pass locally first. See instructions here:
https://docs.openmc.org/en/latest/devguide/tests.html
-->

XinyanBradley commented:
Hi @pshriwise ,
Thank you for reviewing it! It's my pleasure to make contributions to the OpenMC world.
I just noticed that there is a failed C++ format check. Does it matter? How can I fix it?
Xinyan

pshriwise commented:
You should be able to address this by running `clang-format` from the top directory of the repository with 

`clang-format -style=file -i src/particle.cpp`

and then commit and push the resulting changes.

You can also make this step happen automatically in the future by using the `tools/dev/install-commit-hooks.sh` script in the repo. 

I'll also note that we're using `clang-format` version 15 in CI, so you'll want to make sure your local version matches.

XinyanBradley commented:
@pshriwise Thank you, I just did it and it seems it works! Could you see the new commits?
Xinyan

pshriwise commented:
Thanks @XinyanBradley! I just set the CI runs going and yep it looks like that did the trick!

robertsj commented:
I've made a small PR into @XinyanBradley's branch to add checks after multigroup collisions.  A simpler solution overall may be to add a single check in `simulation.cpp` before `p.event_advance()` (or in `event_advance`), but that would make it harder to add the checkpoints @pshriwise suggested.  I started diving into the WW implementation earlier this year and have a couple "sanity check" problems that may make sense for the regression tests, but, at least for a 1-group, slab problem (see an old [class project report](https://github.com/robertsj/iavrtmcsp/blob/master/report/robertsj_project_22106.pdf), the multigroup post-collision check with @XinyanBradley's surface checks produces relative errors consistent with the old report and which are slightly larger than those produced by MCNP 6 for the same problem.  The latter difference is reasonably small and probably due to forced checks for long paths between surfaces and collisions.

robertsj commented on src/particle.cpp:
```diff
@@ -272,6 +273,18 @@ void Particle::event_cross_surface()
   } else {
     // Particle crosses surface
     cross_surface();
+    switch (this->type()) {
```
Would it make sense to move this logic to the `apply_weight_windows` function?

pshriwise commented:
> I've made a small PR into @XinyanBradley's branch to add checks after multigroup collisions. A simpler solution overall may be to add a single check in `simulation.cpp` before `p.event_advance()` (or in `event_advance`), but that would make it harder to add the checkpoints @pshriwise suggested. I started diving into the WW implementation earlier this year and have a couple "sanity check" problems that may make sense for the regression tests, but, at least for a 1-group, slab problem (see an old [class project report](https://github.com/robertsj/iavrtmcsp/blob/master/report/robertsj_project_22106.pdf)), the multigroup post-collision check with @XinyanBradley's surface checks produces relative errors consistent with the old report and which are slightly larger than those produced by MCNP 6 for the same problem. The latter difference is reasonably small and probably due to forced checks for long paths between surfaces and collisions.

Thanks for hammering on the WW features @robertsj! I'd welcome the addition of those regression tests if available. As for where to put the WW checks, I like your suggestion of moving some of that logic into the `apply_weight_windows` function so that adding the call in the right location would enable these checks pretty cleanly. The MFP and mesh element crossings are a little more complicated, but would likely end up in `event_advance` I'd wager.

XinyanBradley commented:
@pshriwise Hello Patrick, per your suggestions, I added the user-specified weight window checkpoints feature. Please let me know if there is anything I could have done better.
BTW there are two commits recently, and only the most recent one, f0bcae, shall be considered. In the previous one e1c97d6, I mistakenly applied clang-format on a python file (settings.py) and it messed up everything.
thank you again for all the supports!
Xinyan

pshriwise reviewed (changes requested):
Thanks for these updates @XinyanBradley! It's looking really good. There are a couple additional steps to tie this up nicely:

  - Addition of test checks that the weight window settings are exported/imported correctly. Ping me if you have questions on where to look for those tests.
  - Documentation of the new parameters in the file I/O sections of the docs (in `/docs` in the repo)

Feel free to take or leave the refactor of the `apply_weight_windows` function. I can tackle that in a follow-up PR.

@paulromano @eepeterson  do you think these are parameters we should export/import from HDF5?

eepeterson commented on openmc/settings.py:
```diff
@@ -945,7 +946,15 @@ def weight_windows_on(self) -> bool:
+    
+    @weight_window_checkpoints.setter
+    def weight_window_checkpoints(self, weight_window_checkpoints: dict):
+        self._weight_window_checkpoints = weight_window_checkpoints
```
nitpick here: rather than assert, something like 
```
if k not in ('collision', 'surface'):
    raise ValueError('useful error message')
````

pshriwise replied:
Yeah good point. That would be better.

pshriwise commented:
> I would say that the hdf5 file should only contain the information about the weight windows themselves, not how they are used. It could be nice to have a fully described weight window implementation in an h5 file and overwrite those values through `openmc.Settings` explictly, but I think that would involve writing more code than it's worth and not having that capability forces people to think about how they are going to use the weight windows, which from my experience is probably a good thing.

Agreed. I think until we see a clear need we can leave this information out of the HDF5 formats for now

XinyanBradley commented:
@eepeterson @pshriwise @robertsj  Thank you so much for your suggestions! The latest commit has the following changes you pointed out:
1. Added type and value input checking for weight_window_checkpoints
2. apply_weight_windows refactorization: moved particle type checks for weight window from Particle::event_cross_surface(Particle& p) and collision(Particle& p) to apply_weight_windows(Particle& p)
3. Documentation of weight_window_checkpoints in /docs/source/io_formats/settings.rst

pshriwise commented on openmc/settings.py:
```diff
@@ -945,7 +946,18 @@ def weight_windows_on(self) -> bool:
+    @weight_window_checkpoints.setter
+    def weight_window_checkpoints(self, weight_window_checkpoints: dict):
+        for key in weight_window_checkpoints.keys():
+            if key not in ('collision', 'surface'):
+                raise ValueError('Wrong values on weight_window_checkpoints. Use either \"collision\" or \"surface\".')
```
@XinyanBradley I should've suggested this the first time. For consistency we should probably use the `check_value` module.
```suggestion
            cv.check_value('weight_window_checkpoint', key, 
            ('collision', 'surface'))
```

XinyanBradley replied:
I will address this in the next commit.

XinyanBradley replied:
Addressed in the latest commit

pshriwise commented on openmc/settings.py:
```diff
@@ -1692,6 +1717,14 @@ def _weight_windows_from_xml_element(self, root, meshes=None):
+        for key in ('collision', 'surface'):
+            value = get_text(elem, key)
+            if value is not None:
+                self.weight_window_checkpoints[key] = value
```
```suggestion
                self.weight_window_checkpoints[key] = value

```

XinyanBradley replied:
Are there any differences?

pshriwise replied:
This adds a line between the end of this function and the one below.

pshriwise commented on src/settings.cpp:
```diff
@@ -967,6 +969,19 @@ void read_settings_xml(pugi::xml_node root)
   }
+
+  // Set up weight window checkpoints
+  if (check_for_node(root, "weight_window_checkpoints")) {
+    xml_node WW_checkpoints = root.child("weight_window_checkpoints");
+    if (check_for_node(WW_checkpoints, "collision")) {
+      weight_window_checkpoint_collision =
+        get_node_value_bool(WW_checkpoints, "collision");
+    }
+    if (check_for_node(WW_checkpoints, "surface")) {
+      weight_window_checkpoint_surface =
+        get_node_value_bool(WW_checkpoints, "surface");
+    }
+  }
```
Sticking to our local variable name conventions here.

```suggestion
  if (check_for_node(root, "weight_window_checkpoints")) {
    xml_node ww_checkpoints = root.child("weight_window_checkpoints");
    if (check_for_node(ww_checkpoints, "collision")) {
      weight_window_checkpoint_collision =
        get_node_value_bool(ww_checkpoints, "collision");
    }
    if (check_for_node(ww_checkpoints, "surface")) {
      weight_window_checkpoint_surface =
        get_node_value_bool(ww_checkpoints, "surface");
    }
  }
```

pshriwise commented on src/weight_windows.cpp:
```diff
@@ -56,6 +56,9 @@ void apply_weight_windows(Particle& p)
   // skip dead or no energy
   if (p.E() <= 0 || !p.alive())
     return;
+  // skip particles other than neutrons and photons
+  if (p.type() != ParticleType::neutron && p.type() != ParticleType::photon)
+    return;
```
This will be handled in `WeightWindows::get_weight_window` when we compare particle types, so I don't think we need it here as well. It does help us avoid allocating a `WeightWindow` object though. Just noting this here
 for posterity I guess. Let's leave it here for now.

XinyanBradley replied:
Addressed in the latest commit

pshriwise commented on src/particle.cpp:
```diff
@@ -272,6 +273,10 @@ void Particle::event_cross_surface()
   } else {
     // Particle crosses surface
     cross_surface();
+    if (settings::weight_windows_on &&
+        settings::weight_window_checkpoint_surface) {
+      apply_weight_windows(*this);
+    }
```
To clean this up further, let's push the check for `settings::weight_windows_on` into `apply_weight_windows` so the first line of that function is:

```cpp
if (!settings::weight_windows_on) return;
```

XinyanBradley replied:
addressed in the latest commit

pshriwise reviewed (changes requested):
Looks excellent @XinyanBradley. Just a few nit picks here and I think this is ready to go!

XinyanBradley commented:
@pshriwise Hi Patrick,
Sorry I forgot to post another comment. I addressed your comments last week. Are there anything else I could have done?
Xinyan

pshriwise commented on src/particle.cpp:
```diff
@@ -272,6 +273,8 @@ void Particle::event_cross_surface()
   } else {
     // Particle crosses surface
     cross_surface();
+    if (settings::weight_window_checkpoint_surface)
```
Pretty clean! Thanks for that refactor!

pshriwise reviewed (changes requested):
Just a couple more items to tidy up and then this should be ready to go! Thanks for tackling this @XinyanBradley!

XinyanBradley commented:
@pshriwise Thank you again for your patient review. All issues are resolved now!

pshriwise reviewed (approved):
Really nice, @XinyanBradley. Thanks for tackling this!
